#summary generalized speciation duplication inference

= Generalized Speciation Duplication Inference =

== Purpose ==

To infer duplication events on a gene tree given a trusted species tree.

== Usage == 
{{{
java -Xmx1024m -cp
path/to/forester.jar org.forester.application.gsdi [-options] <gene tree in phyloXML format> <species tree> <outfile>
}}}

=== Options ===

  * -g: to allow stripping of gene tree nodes without a matching species in the species tree
 
  * -m: use most parimonious duplication model for GSDI: assign nodes as speciations which would otherwise be assiged as potential duplications due tp polytomies in the species tree
  * -q: to allow species tree in other formats than phyloXML (i.e. Newick, NHX, Nexus)
  * -b: to use SDIse algorithm instead of GSDI algorithm (for binary species trees)

==== Gene tree ====
Must be in phyloXM format, with taxonomy and sequence data in appropriate fields.

==== Species tree ====
Must be in phyloXML format unless option -q is used.

=== Options ===
gsdi -g gene_tree.xml tree_of_life.xml out.xml

== Description ==

Currently, GSDI strips the gene tree of nodes for which a matching species is not present in the species tree. 

== Reference ==

Zmasek CM and Eddy SR "A simple algorithm to infer gene duplication and speciation events on a gene tree" [http://bioinformatics.oxfordjournals.org/content/17/9/821.abstract Bioinformatics, 17, 821-828]
 


== Download ==

Download forester.jar here: http://code.google.com/p/forester/downloads/list