#summary generalized speciation duplication inference

= Generalized Speciation Duplication Inference =

== Purpose ==

To infer duplication events on a gene tree given a trusted species tree.

== Usage == 
{{{
java -Xmx1024m -cp
path/to/forester.jar org.forester.application.gsdi [-options] <gene tree in phyloXML format> <species tree> [outfile]
}}}

=== Options ===

  * -s: to strip the species tree prior to duplication inference
  * -b: to use SDI algorithm instead of GSDI algorithm
  * -m: use most parimonious duplication model for GSDI: assign nodes as speciations which would otherwise be assiged as unknown because of polytomies in the species tree
  * -q: to allow species tree in other formats than phyloXML (Newick, NHX, Nexus)

==== Species tree ====
In phyloXML format (unless option -q is used), with taxonomy data in appropriate fields. Must be rooted, polytomies are allowed.

==== Gene tree ====
In phyloXML format, with taxonomy and sequence data in appropriate fields. Must be rooted an binary (no polytomies).


== Description ==

Currently, GSDI strips the gene tree of nodes for which a matching species is not present in the species tree. 

== Reference ==

Zmasek CM and Eddy SR "A simple algorithm to infer gene duplication and speciation events on a gene tree" [http://bioinformatics.oxfordjournals.org/content/17/9/821.abstract Bioinformatics, 17, 821-828]
 


== Download ==

Download forester.jar here: http://code.google.com/p/forester/downloads/list